ROVER-129: Removing Rayon & Crossbeam (Some Of It)

jonathanrainer opened the pull request:
Now that we've made the leap into the world of asynchronous execution we don't need to keep using Rayon thread pools when we could actually get the Tokio runtime to do it for us. This requires a bit of a fiddling to get right, but this should help us overcome the issues we've seen with the smoke tests, and in other places around thread exhaustion with larger numbers of subgraphs.

The refactor is somewhat limited in that we can't remove Crossbeam completely (yet), as it's pretty baked into how Leader/Follower works, however since a refactor is coming down the track there anyway I'm happy to leave that for now. I'd like to do some more manual testing of this against what Fidelity has and make sure we haven't broken anything else but I'm confident this will now work as we expect it to.

Successful smoke test run: https://github.com/apollographql/rover/actions/runs/10596517833/job/29365260226

Have run UAT over this myself and covered the following cases:
- Using a supergraph.yaml with 8 subgraphs ranging from
    - 0 subgraphs with watched schema files and 8 with introspection to
    - 8 subgraphs with introspection and 0 with watched schema files
- Using the Leader/Follower system to spin up 8 `rover dev` sessions with a mix of file watching and introspection
- Using `rover dev --graph-ref`

All of these respond as we would expect! ✅

jonathanrainer committed:
ROVER-129 First tranche of Rayon removed

We're replacing with `tokio` constructs instead

jonathanrainer committed:
ROVER-129 Rayon removed

jonathanrainer committed:
ROVER-129 Remove crossbeam where we can

Disable it for notify (key user), and remove it from rover-std

jonathanrainer committed:
ROVER-129 Fix commentary

jonathanrainer commented on Cargo.toml:
```diff
@@ -102,15 +102,14 @@ interprocess = { version = "2", default-features = false }
 lazycell = "1"
 lazy_static = "1.4"
-notify = "6"
+notify = { version = "6", default-features = false, features = ["macos_kqueue"] }
```
This is due to a known issue with notify: https://github.com/notify-rs/notify/issues/380, we basically turn off crossbeam channels and switch to std ones instead

jonathanrainer reviewed (commented):
Most of the changes are pretty self-explanatory, just one extra note

jonathanrainer commented on src/command/dev/do_dev.rs:
```diff
@@ -120,18 +122,13 @@ impl Dev {
-                });
+            let futs = subgraph_watchers.into_iter().map(|mut watcher| async move {
+                let _ = watcher
+                    .watch_subgraph_for_changes(client_config.retry_period)
+                    .await
+                    .map_err(log_err_and_continue);
             });
-
-            subgraph_watcher_handle
-                .await
-                .expect("could not wait for subgraph watcher thread");
+            tokio::join!(join_all(futs), subgraph_watcher_handle.map(|_| ()));
```
This change stops us spawning tasks that don't actually do anything and will just end up blocking us, behaviour remains the same in testing

loshz commented on src/command/dev/router/config.rs:
```diff
@@ -272,26 +265,23 @@ impl RouterConfigReader {
+                        .recv()
+                        .await
+                        .expect("could not watch router configuration file")
+                        .expect("could not watch router configuration file");
```
Duplicate line above:
```suggestion
```

jonathanrainer replied:
This is an annoying one because it's a change to way `recv()` works when you go `async`. When you call `recv()` now you get back an Option which is None if the channel is closed. So I've changed the idiom here so it's a bit more natural and we don't end up with the duplication.

jonathanrainer committed:
ROVER-129 Fix up reading from channel

Because we're now using `tokio` channels they return None when they
are empty. So we can use a different idiom to read from them,
which ensures this continues to work.

jonathanrainer committed:
Move to using `interval`

This way our Introspection watching works as well
as watching schema files.
